azblob/zt_put_blob_from_url_test.go

```diff
@@ -0,0 +1,302 @@
+	sourceDataMD5Value := md5.Sum(sourceData)
+	ctx := context.Background()
+	srcBlob := container.NewBlockBlobURL(generateBlobName())
+	destBlob := container.NewBlockBlobURL(generateBlobName())
+
+	uploadSrcResp, err := srcBlob.Upload(ctx, r, BlobHTTPHeaders{}, Metadata{}, BlobAccessConditions{}, DefaultAccessTier, nil, ClientProvidedKeyOptions{})
+	c.Assert(err, chk.IsNil)
+	c.Assert(uploadSrcResp.Response().StatusCode, chk.Equals, 201)
+	c.Assert(uploadSrcResp.Response().Header.Get("x-ms-version-id"), chk.NotNil)
+
+	// Get source blob URL with SAS for StageFromURL.
+	srcBlobParts := NewBlobURLParts(srcBlob.URL())
+
+	srcBlobParts.SAS, err = BlobSASSignatureValues{
+		Protocol:      SASProtocolHTTPS,                     // Users MUST use HTTPS (not HTTP)
+		ExpiryTime:    time.Now().UTC().Add(48 * time.Hour), // 48-hours before expiration
+		ContainerName: srcBlobParts.ContainerName,
+		BlobName:      srcBlobParts.BlobName,
+		Permissions:   BlobSASPermissions{Read: true}.String(),
+	}.NewSASQueryParameters(credential)
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	srcBlobURLWithSAS := srcBlobParts.URL()
+
+	// Invoke put blob from URL
+	resp, err := destBlob.PutBlobFromURL(ctx, BlobHTTPHeaders{}, srcBlobURLWithSAS, Metadata{"foo": "bar"}, ModifiedAccessConditions{}, BlobAccessConditions{}, sourceDataMD5Value[:], sourceDataMD5Value[:], DefaultAccessTier, nil, ClientProvidedKeyOptions{})
+	c.Assert(err, chk.IsNil)
+	c.Assert(resp.Response().StatusCode, chk.Equals, 201)
+	c.Assert(resp.Version(), chk.Not(chk.Equals), "")
+	c.Assert(resp.VersionID(), chk.NotNil)
+
+	// Check data integrity through downloading.
+	downloadResp, err := destBlob.BlobURL.Download(ctx, 0, CountToEnd, BlobAccessConditions{}, false, ClientProvidedKeyOptions{})
+	c.Assert(err, chk.IsNil)
+	destData, err := ioutil.ReadAll(downloadResp.Body(RetryReaderOptions{}))
+	c.Assert(err, chk.IsNil)
+	c.Assert(destData, chk.DeepEquals, sourceData)
+	c.Assert(downloadResp.Response().Header.Get("x-ms-version-id"), chk.NotNil)
+	c.Assert(len(downloadResp.NewMetadata()), chk.Equals, 1)
+
+	// Edge case 1: Provide bad MD5 and make sure the put fails
+	_, badMD5 := getRandomDataAndReader(16)
+	_, err = destBlob.PutBlobFromURL(ctx, BlobHTTPHeaders{}, srcBlobURLWithSAS, Metadata{}, ModifiedAccessConditions{}, BlobAccessConditions{}, badMD5, badMD5, DefaultAccessTier, BlobTagsMap{}, ClientProvidedKeyOptions{})
+	c.Assert(err, chk.NotNil)
+
+	// Edge case 2: Not providing any source MD5 should see the CRC getting returned instead
+	resp, err = destBlob.PutBlobFromURL(ctx, BlobHTTPHeaders{}, srcBlobURLWithSAS, Metadata{}, ModifiedAccessConditions{}, BlobAccessConditions{}, nil, nil, DefaultAccessTier, nil, ClientProvidedKeyOptions{})
+	c.Assert(err, chk.IsNil)
+	c.Assert(resp.Response().StatusCode, chk.Equals, 201)
+	c.Assert(resp.Response().Header.Get("x-ms-version"), chk.Equals, ServiceVersion)
+	c.Assert(resp.Response().Header.Get("x-ms-version-id"), chk.NotNil)
```

Comment:
the edge cases seem to be repeated across multiple tests, is this necessary?

Reply:
They are repetitive for the most part so I kept the edge cases on the first test method. Also, for this test, the version ID is being checked in the edge case so I kept this one.